Sources/SwiftDocC/Infrastructure/Link Resolution/ExternalPathHierarchyResolver.swift

```diff
@@ -0,0 +1,235 @@
+            
+            return .failure(unresolvedReference, error.asTopicReferenceResolutionErrorInfo(originalReference: originalReferenceString) { node in 
+                guard let reference = resolvedReference[node.identifier], let summary = content[reference] else {
+                    return node.name
+                }
+                return summary.title
+            })
+        } catch {
+            fatalError("Only PathHierarchy.Error errors are raised from the symbol link resolution code above.")
+        }
+    }
+    
+    private static func path(for unresolved: UnresolvedTopicReference) -> String {
+        guard let fragment = unresolved.fragment else {
+            return unresolved.path
+        }
+        return "\(unresolved.path)#\(urlReadableFragment(fragment))"
+    }
+
+    /// Returns the external entity for a symbol's unique identifier or `nil` if that symbol isn't known in this external context.
+    func entity(symbolID usr: String) -> ExternalEntity? {
+        // TODO: Resolve external symbols by USR (rdar://116085974) (There is nothing calling this function)
+        // This function has an optional return value since it's not easy to check what module a symbol belongs to based on its identifier.
+        guard let reference = symbols[usr] else { return nil }
+        return entity(reference)
+    }
+    
+    /// Returns the external entity for a reference that was successfully resolved by this external resolver.
+    ///
+    /// - Important: Passing a resolved reference that wasn't resolved by this resolver will result in a fatal error.
+    func entity(_ reference: ResolvedTopicReference) -> ExternalEntity {
+        guard let resolvedInformation = content[reference] else {
+            fatalError("The resolver should only be asked for entities that it resolved.")
+        }
+        
+        let topicReferences: [ResolvedTopicReference] = (resolvedInformation.references ?? []).compactMap {
+            guard let renderReference = $0 as? TopicRenderReference,
+                  let url = URL(string: renderReference.identifier.identifier),
+                  let bundleID = url.host
+            else {
+                return nil
+            }
+            return ResolvedTopicReference(bundleIdentifier: bundleID, path: url.path, fragment: url.fragment, sourceLanguage: .swift)
+        }
+        let dependencies = RenderReferenceDependencies(
+            topicReferences: topicReferences,
+            linkReferences: (resolvedInformation.references ?? []).compactMap { $0 as? LinkReference },
+            imageReferences: (resolvedInformation.references ?? []).compactMap { $0 as? ImageReference }
+        )
+        
+        return .init(
+            topicRenderReference: resolvedInformation.topicRenderReference(),
+            renderReferenceDependencies: dependencies,
+            sourceLanguages: resolvedInformation.availableLanguages
+        )
+    }
+    
+    // MARK: Deserialization
```

Comment:
These files should be small compared to symbol graph files and they are being decoded while we wait for the symbol graph files to be processed and markdown files to be parsed. 

That said, I haven't measured this for a long time — since some early prototypes — so I should do that again.

Reply:
I did a number of different benchmarks on this today and the hardest part is knowing what to compare...

A "size 10" test bundle decoding another "size 10" test bundle's link files is ~1.5-2% slower total convert time, which is around what I'd expect. 

However, emitting the link files compared to not emitting the link file is a 10% slower convert time, which I didn't expect. I looked into this some more and noticed that emitting the digest files is 15% slower than not emitting them, so it's looks like most of the slowdown comes from computing the linkable entities. That's definitely something that we should continue looking at but I don't feel like it should be addressed in this PR.

Reply:
I filed rdar://116495939 to continue investigating the performance impact of resolving links to content in other archives.

Reply:
Sounds good. Thank you!